[net9.0] Merge main to net9.0

rmarinho opened the pull request:
### Description of Change

Bring latest changes from main to net9.0

Commit:
fix(collectionview): Fixed scroll to group item on Android + Windows

Fixed scrolling to a group index not working on Windows
Implemented scroll to group index on Android#8718

Commit:
Add -1 check for group index

Foda committed:
Merge branch 'main' into foda/ScrollToGroup

Commit:
Merge branch 'main' into foda/ScrollToGroup

# Conflicts:
#	src/Controls/tests/DeviceTests/Elements/CollectionView/CollectionViewTests.cs

Commit:
Resolved conflicts

PureWeen committed:
Merge branch 'main' into foda/ScrollToGroup

daltzctr committed:
Adopt chromium methodology for determining tablet on Windows

daltzctr committed:
Fix auto rotation state enums being incorrect

daltzctr committed:
Fix bitfield comparison

daltzctr committed:
Remove unused const, refactor comments

Commit:
Merge branch 'main' into foda/ScrollToGroup

Commit:
Fix test

Commit:
Skip test on ios/mac

jfversluis committed:
Fix HostIdentifier bindings for item templates (#23912)

rmarinho committed:
[ci] Revert arcade version bump (#23930)

MartyIX committed:
[Windows] Handle non-seekable streams in `PlatformImage.FromStream` (take 2) (#23824)

* Fix

* Optimize

* Seek back

* Optimization

* Use Microsoft.IO.RecyclableMemoryStream

* Not needed?

* It seems I can remove version now

* Review

* Add test

PureWeen committed:
Update Versions for bug reports

PureWeen committed:
Add SR7 Versions for bug reports (#23990)

mattleibow committed:
Use explicit platform versions (#23995)

This is used to avoid the case where some SDKs bump their platform
version and we end up forcing the entire ecosystem to update SDKs.

mattleibow committed:
Remove the 'Resources' prefix from BundleResource (#23978)

* Remove the 'Resources' prefix from BundleResource

Fixes #23554

Prior to Maui 8.0.70, there was a bug that only removed the "Resources/" prefix, and not the correct "Platforms/iOS/Resource": https://github.com/dotnet/maui/issues/16734

This PR https://github.com/dotnet/maui/pull/23269 fixes the original issue, but now exposed the case where BundleResource were included in the root Resources folder instead of the Platforms/iOS/ folder.

* Update Microsoft.Maui.Controls.SingleProject.targets

* Update Microsoft.Maui.Controls.SingleProject.targets

* Add a test

* string

albyrock87 committed:
Fixes Android FontModel Glide In-Memory Caching (#24021)

PureWeen committed:
[iOS] Add a Container around ImageButton (#24020)

* Add a Container around ImageButton

* Update PublicAPI.Unshipped.txt

* - don't animate push

jfversluis committed:
Update Rebase GitHub Action (#24031)

Update checkout step to use checkout at v4 which uses a newer Node version. v2 uses a Node version that might be removed at some point.

Foda committed:
[Windows] Fix CarouselView not showing new items (#23779)

* Fix missing vector change event propigation

* Add test

* Add ref images
Add WeakVectorChangedChangedProxy

---------

Co-authored-by: Mike Corsaro <mikecorsaro@microsoft.com>

jpiechowiak committed:
Added condition in WindowStateManager for selecting proper scene used with role Application (#23945)

PureWeen committed:
Check to see if Image Should Reload When Attached (#24023)

* Check to see if Image Should Reload When Attached

* - cleanup

* Update ImageSourcePartLoader.cs

* - fix images

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/xharness build 20240729.2 (#24012)

Microsoft.DotNet.XHarness.CLI , Microsoft.DotNet.XHarness.TestRunners.Common , Microsoft.DotNet.XHarness.TestRunners.Xunit
 From Version 9.0.0-prerelease.24374.1 -> To Version 9.0.0-prerelease.24379.2

Co-authored-by: dotnet-maestro[bot] <dotnet-maestro[bot]@users.noreply.github.com>

jsuarezruiz committed:
[UI Testing] Add legacy tests (alternative) (#22635)

* Added Compatibility UITests (ported to Appium) to TestCases project

* Added tests pages

* More changes

* More changes

* More changes

* More changes

* More fixes

* More fixes

* More changes

* More changes

* More changes

* More changes

* More fixes

* Avoid problems with duplicated issues

* More changes

* More changes

* More fixes

* Fixes on Windows tests

* More fixes

* More fixes

* More changes

* Removed duplicated tests

* More fixes

* More changes

* More changes

* More changes

* More changes

* More changes

* Removed unnecessary changes

* More changes

* Ignore golden tests on macOS

* More changes

* More changes

* More changes

* Changes in AppiumLifecycleActions

* More changes

* Removed IgnoreIfPlatforms usage

* Small changes

* Fix build errors

* Fix build error

* Add missing category

* Update Issue24574.cs

* Ignore on Mac for now

---------

Co-authored-by: Shane Neuville <shneuvil@microsoft.com>
Co-authored-by: Gerald Versluis <gerald.versluis@microsoft.com>

mattleibow committed:
Update the aar from CI (#24038)

PureWeen committed:
Bump main to 8.0.90 (#24050)

kubaflo committed:
[Android] Border Stroke gradient can only switch to another gradient - fix (#22168)

* Fixed the issue #22042 & added a UITest

* Added snapshot

* Added a test category

* Updated a Ui test

* Update Issue22042.cs

* Added snapshots

* Fixed WIndows test

* Windows snapshot

---------

Co-authored-by: Javier Suárez <javiersuarezruiz@hotmail.com>

PureWeen committed:
Update bug-report.yml with 8.0.80 SR8 (#24083)

kubaflo committed:
[Testing] Added a support for the device's theme change (#22508)

* [Testing] Added a support for the device's theme change

* Comments resolved

* Use Adb

* UI Test improvement

* Updated test

* More snapshots

* [Testing] Added a support for the device's theme change

* Comments resolved

* Use Adb

* UI Test improvement

* Updated test

* More snapshots

* Update ThemeChange.cs

* Test fixes

* Added the first android snapshot

* Create AppThemeShouldChangeLightTheme.png

---------

Co-authored-by: Javier Suárez <javiersuarezruiz@hotmail.com>
Co-authored-by: Jonathan Dick <jodick@microsoft.com>
Co-authored-by: Gerald Versluis <gerald.versluis@microsoft.com>

MartyIX committed:
[Windows][GestureManager] Unsubscribe control tap events only if they were subscribed (#23976)

* Optimize

* feedback

* More

* fix build

mattleibow committed:
Merge branch 'main' into dev/merge-main-net9

jfversluis committed:
Remove excessive Grid API comment (#24037)

jonathanpeppers committed:
[android] use cached value for `GetDisplayDensity()` (#23666)

Context: https://github.com/davidortinau/AllTheLists

Profiling @davidortinau's app, I noticed the "Check-ins" sample felt
the slowest on Android.

One thing I noticed while scrolling:

    44.37ms (0.48%) Microsoft.Maui!Microsoft.Maui.Platform.StrokeExtensions.UpdateMauiDrawable(Android.Views.View,Microsoft.Maui.IBorderStr
    40.59ms (0.44%) Microsoft.Maui!Microsoft.Maui.Graphics.MauiDrawable..ctor(Android.Content.Context)

Where a lot of the time is actually just spent in:

    54.48ms (0.59%) Microsoft.Maui!Microsoft.Maui.Platform.ContextExtensions.GetDisplayDensity(Android.Content.Context)

The `GetDisplayDensity()` method is called a lot, and it's not using
the cached value we already have!

Reading the history in 6babd4ec, there does not appear to be a reason
we *didn't* use the cached value.

With this change in place, I see instead:

    7.81ms (0.12%) Microsoft.Maui!Microsoft.Maui.Platform.StrokeExtensions.UpdateMauiDrawable(Android.Views.View,Microsoft.Maui.IBorderStr

And I don't see *any* time spent in `MauiDrawable..ctor()` anymore.

This change should improve the performance of any `<Border/>` on Android.

jonathanpeppers committed:
[xaml] improve performance of `BindableProperty.TryConvert` (#23685)

* [xaml] improve performance of `BindableProperty.TryConvert`

Context: https://github.com/davidortinau/AllTheLists

Profiling @davidortinau's app, I noticed the "Check-ins" sample felt
the slowest on Android.

One thing I noticed while scrolling:

    58.46ms (0.87%) Microsoft.Maui.Controls!Microsoft.Maui.Controls.BindableProperty.TryConvert(object&)
    47.96ms (0.74%) Microsoft.Maui.Controls!Microsoft.Maui.Controls.Xaml.TypeConversionExtensions.GetImplicitConversionOperator(System.Type,System.Type,System.Type)

`GetImplicitConversionOperator()` is using System.Reflection to look up
if a type has an implicit conversion operator. This is a slow operation.

I added some logging, to find out which types trigger this:

    System.String -> Microsoft.Maui.Controls.ImageSource
    System.Double -> System.String
    System.String -> System.Double

For the first, `ImageSource`, we can add an entry to
`KnownTypeConverters`.

For the second, `System.Double -> System.String`, we can add an entry
to `SimpleConvertTypes`. I also added an entry for `System.Bool`.

Lastly, I don't think we should actually be looking up implicit
operators on built-in, primitive types like: `double`, `string`,
`bool`, etc. I added a check to skip these.

With this change in-place, `GetImplicitConversionOperator()` is no
longer called anymore in this app:

    12.29ms (0.23%) Microsoft.Maui.Controls!Microsoft.Maui.Controls.BindableProperty.TryConvert(object&)

This should improve the performance of data-binding on all platforms.

This should hopefully not change any behavior and simply avoids
looking up implicit operators on types that do not contain any.

* Fix test failures

jfversluis committed:
Remove excessive StackLayout API docs (#24114)

jfversluis committed:
Add shared project to primaryOutputs (#24095)

Foda committed:
Ensure border stroke does not regenerate on Parent or Window property changes (#24125)

Co-authored-by: Mike Corsaro <mikecorsaro@microsoft.com>

Eilon committed:
Remove instances of '#if NET6_0_OR_GREATER' (#24126)

We're way past that, so deleting these unused cases.

MartyIX committed:
Allow setting `PlatformCanvas.Antialias` on Windows (#23965)

mattleibow committed:
Merge remote-tracking branch 'origin/net9.0' into merge-main-net9

kubaflo committed:
[Android] App crash when using a binding value that exceeds Entry max length (#24138)

mattleibow commented:
/azp run

jfversluis committed:
Remove MAUI styling from Blazor Hybrid template (#24135)

* Remove MAUI styling from Blazor Hybrid template

* Use fwlink

mattleibow committed:
Merge remote-tracking branch 'origin/net9.0' into merge-main-net9

mattleibow committed:
Merge remote-tracking branch 'origin/main' into merge-main-net9

# Conflicts:
#	src/Controls/src/Core/BindableProperty.cs

PureWeen committed:
Move EnsureSupportForCustomWebViewClients to Appium (#24167)

* Move EnsureSupportForCustomWebViewClients to Appium

* - fix

Foda committed:
Fix memory leak w/ main window (#24176)

Co-authored-by: Mike Corsaro <mikecorsaro@microsoft.com>

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/xharness build 20240805.1 (#24154)

Microsoft.DotNet.XHarness.CLI , Microsoft.DotNet.XHarness.TestRunners.Common , Microsoft.DotNet.XHarness.TestRunners.Xunit
 From Version 9.0.0-prerelease.24379.2 -> To Version 9.0.0-prerelease.24405.1

Co-authored-by: dotnet-maestro[bot] <dotnet-maestro[bot]@users.noreply.github.com>

jsuarezruiz committed:
Fix Shell.NavBarIsVisible property (#20173)

* Fix the issue

* Added XAML Test

tj-devel709 committed:
make flyout titles on iOS read as buttons (#24018)

Eilon committed:
Add xUnit types to avoid first-chance exceptions in tests (#24130)

* WIP: Add xUnit types to avoid first-chance exceptions in tests

While debugging tests if you have the debugger set to catch *all* exceptions, you'll see a bunch of them for various xUnit types, such as:

System.MissingMethodException: 'Constructor on type 'Xunit.Sdk.InlineDataDiscoverer' not found.'

While they're harmless from a functional perspective, several of these happen when you start debugging tests.

This change re-implements the affected xUnit types as well as a custom MAUI test type to avoid the exception.

* Update xUnitCustomizations.cs

* Clean up code

Foda committed:
Fixed scroll to group item on Android + Windows (#16704)

### Description of Change

Fixed scrolling to a group via `CollectionView.ScrollTo(index,
groupIndex)` not working on Windows or Android.

### Issues Fixed

Fixes #8718


### Demo

The demo below shows scrolling to the group letter "N"


https://github.com/dotnet/maui/assets/890772/af943949-e10d-416e-b928-d7114f1a48b7


https://github.com/dotnet/maui/assets/890772/51da9dac-f18a-4894-865a-30273433b0b4

jfversluis committed:
Remove reference to removed styling (#24199)

mattleibow committed:
Merge branch 'net9.0' into merge-main-net9

mattleibow committed:
Merge branch 'main' into merge-main-net9

# Conflicts:
#	eng/Version.Details.xml

TommiGustafsson-HMP committed:
Fixed low quality images in Resizetizer (#23924)

* Fixed low quality images in Resizetizer, issue #23212.

* Regenerate all images

---------

Co-authored-by: Matthew Leibowitz <mattleibow@live.com>

mattleibow commented:
The `CollectionScrollToGroupWorks` test is new and failing on main. Unrelated to this PR.

mattleibow committed:
Merge remote-tracking branch 'origin/main' into merge-main-net9

mattleibow committed:
Merge remote-tracking branch 'origin/net9.0' into merge-main-net9

mattleibow committed:
Disable 2 failing tests #24243

These are now failing after merging main into net9.0
https://github.com/dotnet/maui/issues/24243

jsuarezruiz committed:
[Testing] Added specific actions by platform (#21330)

* Added platform specific Android Appium actions

* Adding inline comments

* More methods

* iOS specifics

* Changs based on PR feedback

* Fix test

* Update Issue18740.cs

---------

Co-authored-by: Gerald Versluis <gerald.versluis@microsoft.com>

jsuarezruiz committed:
[Testing] Added UITest to validate the Switch OnColor property (#20542)

* Added UITest to validate the Switch OnColor property

* Added snapshots

* Updated snapshot

* Updated snapshots

* Updated snapshot

* Update Issue20535.cs

* Add Category

* Move Windows snapshot

* Move snapshot to the actual right place

---------

Co-authored-by: Gerald Versluis <gerald.versluis@microsoft.com>

jonathanpeppers committed:
[android] call `PlatformContentViewGroup.postInvalidate()` less (#24222)

Context: https://github.com/dotnet/maui/issues/23991

@chabiss's sample app is running a lot of GCs:

    [riodictablemaui] Explicit concurrent copying GC freed 933(127KB) AllocSpace objects, 31(4092KB) LOS objects, 49% free, 4423KB/8846KB, paused 1.328ms,169us total 33.513ms
    [riodictablemaui] Explicit concurrent copying GC freed 722(46KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 4581KB/9162KB, paused 1.856ms,170us total 14.835ms
    [riodictablemaui] Explicit concurrent copying GC freed 641(43KB) AllocSpace objects, 1(108KB) LOS objects, 49% free, 4525KB/9050KB, paused 1.420ms,176us total 14.112ms
    [riodictablemaui] Explicit concurrent copying GC freed 619(41KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 4579KB/9158KB, paused 1.263ms,173us total 13.825ms
    ... many many more lines

This doesn't seem like a normal amount of GCs for a simple app like this.

Attaching Android Studio's allocation profiler, I can see a significant
amount of allocations at:

    com.microsoft.maui.PlatformContentViewGroup.setHasClip()
        Allocations: 35,906
        Deallocations: 143
        Total Count: 35,763
        Shallow Size: 2,288,832

The call to `postInvalidate()` is running *a lot*, and this causes:

* A layout cycle

* `android.os.Message` to be allocated

* etc. etc.

I think we can simply check if `hasClip` has changed, and only call
`postInvalidate()` if it has.

Additionally, we can apply same change for `WrapperView.setHasShadow()`.

kubaflo committed:
NavigationBar properties not updating - fix (#23076)

* Title not updated after OnAppearing for TabbedPage - fix (#22899)

* Added a test category

* Update Issues22899.xaml

* Added snapshots

Foda committed:
Adopt chromium methodology for determining tablet on Windows (#22809)

### Description of Change

Uses a more conservative approach of determining tablet mode on Windows.
This approach is functionally identical to [Chromium's implementation](

https://source.chromium.org/chromium/chromium/src/+/main:base/win/win_util.cc;l=537;drc=ac83a5a2d3c04763d86ce16d92f3904cc9566d3a;bpv=0;bpt=1)
for determining tablet on Windows. Additionally, it adds a few
documentation comments on the observed behavior.

I don't have a good idea on how to add a unit test for this, and it's a
bit of an implicit breaking change (some users who had tablet layouts
may have desktop layouts now), but I think that number of users affected
will be near, or at 0.

### Issues Fixed

Possibly fixes #22093

jfversluis committed:
Update bug report template with .NET 9 preview 7 (#24261)

mattleibow committed:
Merge remote-tracking branch 'origin/main' into merge-main-net9

# Conflicts:
#	src/Core/src/maui.aar

PureWeen committed:
Remove Failing Tests for Now

PureWeen commented:
- RunOniOS failures are not related to this PR
